src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs

```diff
@@ -151,6 +155,61 @@ static BoundNode unwrapListElement(BoundCollectionExpression node, BoundNode ele
             }
         }
 
+        // If we have something like `List<int> l = [.. someEnumerable]`
+        // try rewrite it using `Enumerable.ToList` member if possible
+        private bool TryRewriteSingleElementSpreadToList(BoundCollectionExpression node, TypeWithAnnotations listElementType, [NotNullWhen(true)] out BoundExpression? result)
+        {
+            result = null;
+
+            if (node.Elements is not [BoundCollectionExpressionSpreadElement singleSpread])
+            {
+                return false;
+            }
+
+            if (!ShouldUseAddRangeOrToListMethod(singleSpread.Expression.Type, singleSpread.EnumeratorInfoOpt?.GetEnumeratorInfo.Method, listElementType))
+            {
+                return false;
+            }
+
+            if (TryGetWellKnownTypeMember(node.Syntax, WellKnownMember.System_Linq_Enumerable__ToList, out MethodSymbol? toListGeneric, isOptional: true))
+            {
+                var toListOfElementType = toListGeneric.Construct([listElementType]);
+                result = _factory.Call(receiver: null, toListOfElementType, singleSpread.Expression);
+                return true;
+            }
+
+            return false;
+        }
+
+        private bool ShouldUseAddRangeOrToListMethod(TypeSymbol? spreadType, MethodSymbol? getEnumeratorMethod, TypeWithAnnotations elementType)
+        {
+            Debug.Assert(spreadType is not null);
+
+            var iEnumerableOfTType = _compilation.GetSpecialType(SpecialType.System_Collections_Generic_IEnumerable_T);
+            var iEnumerableOfElementType = iEnumerableOfTType.Construct([elementType]);
+
+            var discardedUseSiteInfo = CompoundUseSiteInfo<AssemblySymbol>.Discarded;
+
+            var conversion = _compilation.Conversions.ClassifyImplicitConversionFromType(spreadType, iEnumerableOfElementType, ref discardedUseSiteInfo);
```

Comment:
Should we maybe use the actual use site info for thisand report it if we take the optimzal code path an the end? Unline the `ICollection<T>` check, the resulting conversion is kinda used in the final codegn, although implicitly

Reply:
I guess the original code for the AddRange case was reporting the use-site diagnostic, but I think it's fine to drop it. The code we generate does not actually use this type in any way, it's just a hint for us to decide if AddRange/ToList is likely to be efficient.